### PR TITLE
Refactor

### PR DESCRIPTION
**Pull request by mgates3**
Previously, the BLAS wrappers are a single layer: `blas::gemm` has 4 implementations (s, d, c, z) that look basically the same and directly call the Fortran BLAS. This refactors it into 3 layers:
1. Low-level wrappers `internal::gemm` are overloaded calls to Fortran BLAS.
2. Mid-level wrapper `impl::gemm` is templated and has all the BLAS++ logic to check arguments, deal with row/col layout, etc.
3. High-level wrappers `blas::gemm` are overloaded calls for float, double, complex-float, complex-double.

The goal is to have only one implementation in (2), to reduce replication of BLAS++ logic code. SLATE's ScaLAPACK wrappers use a similar layered approach.

[I committed gemm separate from the others to serve as an example](https://github.com/icl-utk-edu/blaspp/pull/18/commits/047e5eda52327c42e6925e02e5f5c1b6909705fc). Basically it replaces 1 layer:
```
void gemm( ..., float*  A, ... ) { ...BLAS++ logic...; BLAS_sgemm( ... ); }
void gemm( ..., double* A, ... ) { ...BLAS++ logic...; BLAS_dgemm( ... ); }
void gemm( ..., complex<float>*  A, ... ) { ...BLAS++ logic...; BLAS_cgemm( ... ); }
void gemm( ..., complex<double>* A, ... ) { ...BLAS++ logic...; BLAS_zgemm( ... ); }
```
with 3 layers:
```
// Low-level calls Fortran
internal {
void gemm( ..., float*  A, ... ) { BLAS_sgemm( ... ); }
void gemm( ..., double* A, ... ) { BLAS_dgemm( ... ); }
void gemm( ..., complex<float>*  A, ... ) { BLAS_cgemm( ... ); }
void gemm( ..., complex<double>* A, ... ) { BLAS_zgemm( ... ); }
}

// Mid-level checks & converts arguments, deals with layout, calls low-level.
impl {
template <typename scalar_t>
void gemm( ..., scalar_t* A, ... ) { ...BLAS++ logic...; internal::gemm( ... ); }
}

// High-level calls mid-level.
void gemm( ..., float*  A, ... ) { impl::gemm( ... ); }
void gemm( ..., double* A, ... ) { impl::gemm( ... ); }
void gemm( ..., complex<float>*  A, ... ) { impl::gemm( ... ); }
void gemm( ..., complex<double>* A, ... ) { impl::gemm( ... ); }
```

There's also a separate templated `blas::gemm` in [include/blas/gemm.hh](https://github.com/icl-utk-edu/blaspp/blob/master/include/blas/gemm.hh), not part of this PR, that is a generic implementation of gemm, i.e., it doesn't call Fortran BLAS, it is just the naïve triple-nested loop implementation. Similarly for other routines. This is most helpful for Level 1 & 2 BLAS, where the naïve implementation is okay; less helpful for Level 3 BLAS, which needs extensive optimization.

Why not just stop with (2)? Besides the templated implementation mentioned above, templates are also finicky about exactly matching their arguments. For instance:
```
    float *A, *B, *C;
    blas::gemm( ..., -1.0, A, lda, B, ldb, 1.0, C, ldc );    // works with overload, doesn't work with template
    blas::gemm( ..., -1.0f, A, lda, B, ldb, 1.0f, C, ldc );  // works with both
```
The first call works if `blas::gemm` is overloaded as in (3), but won't compile if `blas::gemm` is templated as in (2) because it doesn't know whether to make scalar_t = double or float. Similarly:
```
    std::complex<double> *A, *B, *C;
    blas::gemm( ..., -1.0, A, lda, B, ldb, 1.0, C, ldc );  // works with overload, not with template
    blas::gemm( ..., std::complex<double>( -1.0 ), A, lda, B, ldb, std::complex<double>( 1.0 ), C, ldc );  // works with both
```

### CODE REVIEW

File: src/dot.cc [hunk 1]
```diff
@@ -5,246 +5,293 @@
...
         // GNU gcc convention
         blas_complex_float value
-             = BLAS_cdotc( &n_,
-                           (blas_complex_float*) x, &incx_,
-                           (blas_complex_float*) y, &incy_ );
+             = BLAS_cdotc( &n,
+                           (blas_complex_float*) x, &incx,
+                           (blas_complex_float*) y, &incy );
         return *reinterpret_cast< std::complex<float>* >( &value );
     #endif
```
Comment (cayrols) [inline src/dot.cc:61]: This structure is duplicated multiple times. Could we make a wrapper? Like:
```
template <scalar_t, blas_t, f>
scalar_t dot( n, x, incx, y, incy ) {
//Contains the ifdef and reinterpret_Cast, and call the function f
[...]
}
```

File: src/ger.cc [hunk 2]
```diff
@@ -130,58 +150,48 @@ void ger(
...
     }
 }
 
-// -----------------------------------------------------------------------------
-/// @ingroup ger
-void ger(
+//------------------------------------------------------------------------------
+/// Mid-level templated wrapper checks and converts arguments,
+/// then calls low-level wrapper.
+/// Unonjugated x y^T version.
```
Comment (cayrols) [inline src/ger.cc:184]: unconjugated

File: src/ger.cc [hunk 3]
```diff
@@ -5,118 +5,138 @@
...
-/// @ingroup ger
-void ger(
-    blas::Layout layout,
-    int64_t m, int64_t n,
-    float alpha,
-    float const *x, int64_t incx,
-    float const *y, int64_t incy,
-    float       *A, int64_t lda )
+//------------------------------------------------------------------------------
+/// Low-level overload wrapper calls Fortran, float version.
```
Comment (cayrols) [inline src/ger.cc:18]: Unless I misunderstand it, I think this says it is the `float` version, but the code is considering `std::complex<float>`.
But I have the feeling it does not exist maybe.
Anyhow, I would say it is missing the `float` and `double` version of `ger` or the comments should state `std::complex<float>` as it is for `geru`.

File: src/her.cc [hunk 4]
```diff
@@ -58,99 +74,99 @@ void her(
...
     }
-
     char uplo_ = uplo2char( uplo );
-    BLAS_cher( &uplo_, &n_,
-               &alpha,
-               (blas_complex_float*) x2, &incx_,
-               (blas_complex_float*) A, &lda_ );
+
+    // call low-level wrapper
+    internal::her( uplo_, n_,
```
Comment (cayrols) [inline src/her.cc:101]: Just thinking. Should not it be like:
```
internal::her(
  uplo_, [....] );
```

File: src/iamax.cc [hunk 5]
```diff
@@ -5,98 +5,128 @@
...
 {
-    // check arguments
-    blas_error_if( n < 0 );      // standard BLAS returns, doesn't fail
-    blas_error_if( incx <= 0 );  // standard BLAS returns, doesn't fail
+    return BLAS_isamax( &n, x, &incx );
+}
+
+//------------------------------------------------------------------------------
+/// Low-level overload wrapper calls Fortran, double version.
+/// Low-level overload wrapper calls Fortran, double version.
```
Comment (cayrols) [inline src/iamax.cc:29]: Twice.

File: src/ger.cc (hunk 3 above)
Comment (mgates3) [reply src/ger.cc:18]: Fixed.

File: src/dot.cc (hunk 1 above)
Comment (cayrols) [reply src/dot.cc:61]: I see. It might require too much effort for now. But I thought, by passing `f` as a template it would work.

**Review by cayrols — APPROVED** [review]
I think it is good and ready to be merged.

File: src/her.cc (hunk 4 above)
Comment (mgates3) [reply src/her.cc:101]: The style guide allows either
```
foobar( a, b, c,
        x, y, z );
```
or
```
foobar(
    a, b, c,
    x, y, z );
```
if that's what you're asking about. For prototypes, I prefer the later, but for calls I use either one.

**Pull request merged**